OurDNA: Dashboard Backend

nevoodoo opened the pull request:
## Description

This PR adds the backend logic that aggregates the data needed for the OurDNA Dashboard UI.

See: https://docs.google.com/spreadsheets/d/1crEwZhFMD-m0sO4TGm-8FcwsFxGhg_IIyPB5yH2O7Q0/edit?usp=sharing

## Changes
- Added initial layer logic at `db/python/layers/ourdna/dashboard.py` that controls the aggregation logic.
- Added initial schema changes to `api/graphql/schema.py` that would offer a new field to query this data from.

illusional commented on api/graphql/schema.py:
```diff
@@ -1074,6 +1075,31 @@ async def analysis_runner(
+        if not project:
+            raise ValueError('Must provide project')
+
+        project_name_map: dict[str, int] = {}
+        if project:
+            project_names = project.all_values()
+            projects = await ptable.get_and_check_access_to_projects_for_names(
+                user=connection.author, project_names=project_names, readonly=True
+            )
+            project_name_map = {p.name: p.id for p in projects}
+
+        filter_ = SampleFilter(
+            project=(
+                project.to_internal_filter(lambda pname: project_name_map[pname])
+                if project
+                else None
+            ),
+        )
```
`project` is always going to have a value here, so you can skip the following nullable checks. Also, do you want this to be multiple projects, or just a single?

vivbak replied:
We were thinking the project here would just be 'OurDNA'. Is there a scenario where another project is valid here?

illusional commented on api/graphql/schema.py:
```diff
@@ -1074,6 +1075,31 @@ async def analysis_runner(
+                else None
+            ),
+        )
+        return [await OurDnaDashboardLayer(connection).query(filter_)]
```
This reads a bit awkward, do you actually want to return multiple values here?

nevoodoo replied:
Mmmm, GraphQL refused to accept anything but an iterable here :(

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -0,0 +1,140 @@
+
+        self.sample_table = SampleTable(connection)
+
+    async def get_collection_to_process_end_time(self, sample: Sample) -> int | None:
```
FWIW, if you don't expect this to be nullable, I'd probably not mark it as so. It can give a false illusional it's making a async check.

Also, definitely not expecting these to already be here, but these methods are very test-able. Also consider the cases where the values are not in the expected format.

nevoodoo replied:
I think where `process_end_time` is not available yet because the sample isn't processed, or it is lost, we wouldn't be able to make this calculation. And so we have two cases here to be handled: not processed/lost OR processed. Thoughts?

illusional replied:
Yeah totally agree, just make sure it's handled as strptime will raise a ValueError.
(I meant marking this as async, not nullable, got my words all mixed up)

nevoodoo replied:
Removed the async :) I think the handling should be working fine but will include this check in testing later.

illusional reviewed (commented):
Looking great so far! I don't know how this is recorded in the user stories, but you may want to think carefully about the relationship between samples + participant stats, and make sure any language around it is super clear and we're trying to display what's actually important to someone.

EG: _90% of samples processed within 24 hours_ may be deceiving if we split samples further if it's delivered under 24 hours, hence it may only be 20% of participant's have a sample that was processed in 24 hours.

nevoodoo commented:
@vivbak @illusional I have moved the `ourdna` field to the `GraphQLProject` type, and removed the nullable checks like was suggested, and there is a project permissions check being made by the layer so I have removed it from the schema level for now. Keen to hear what you think, I have tested to make sure the query works as intended locally :)

illusional commented on api/graphql/schema.py:
```diff
@@ -205,6 +206,16 @@ async def analysis_runner(
+        self,
+        info: Info,
+        root: 'Project'
+    ) -> list[strawberry.scalars.JSON]:
```
```suggestion
    ) -> strawberry.scalars.JSON:
```

illusional commented on api/graphql/schema.py:
```diff
@@ -205,6 +206,16 @@ async def analysis_runner(
+    ) -> list[strawberry.scalars.JSON]:
+        connection = info.context['connection']
+        ourdna_layer = OurDnaDashboardLayer(connection)
+        return [await ourdna_layer.query(SampleFilter(project=root.id), project_id=root.id)]
```
```suggestion
        return await ourdna_layer.query(SampleFilter(project=root.id), project_id=root.id)
```

vivbak commented:
To Action: 
https://github.com/populationgenomics/metamist/pull/766#discussion_r1604073456

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -0,0 +1,269 @@
+        Get a property from the meta field of a sample, accounting for hyphenated property names
+        or underscores in the property name
+        """
+        return sample.meta.get(property_name) or sample.meta.get(property_name.replace('-', '_'))
```
Are these lines not formatted to 88, they feel longer?

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -0,0 +1,269 @@
+        participants_consented_not_collected: list[int] = []
+        participants_signed_not_consented: list[int] = []
+
+        # TODO We should figure out if we need to handle more than a single project_id
```
I don't think you need to handle more than a single project here, at least not for now

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -0,0 +1,269 @@
+        self,
+        filter_: SampleFilter,
+        check_project_ids: bool = True,
+        project_id: int = None,
```
This parameter is 
```suggestion
        project_id: ProjectId = None,
```

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -0,0 +1,269 @@
+        project_id: int = None,
+    ) -> dict:
+        """Get dashboard data"""
+        projects: set[int]
```
```suggestion
        projects: set[ProjectId]
```

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -0,0 +1,269 @@
+            await self.participant_table.get_participants(project=project_id)
+        )
+
+        projects, samples_internal = await self.sample_table.query(filter_=filter_)
```
Call the SampleLayer query method, and then you don't need to check projects

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -0,0 +1,269 @@
+        # TODO We should figure out if we need to handle more than a single project_id
+        # for project_id in project_ids:
+        participants.extend(
+            await self.participant_table.get_participants(project=project_id)
```
- Why do you take a sample filter, and get all participants? Do you just want participants for the fetched samples, or actually want all. Can you add a comment either way.
- (pending the previous dot point): you can run both the participant and sample query together, in an `asyncio.gather`.
- Better to call a participant layer's `query` method, rather than interact with tables if you can avoid

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -0,0 +1,269 @@
+            )
+
+        for participant in participants:
+            samples_for_participant = [
```
This is an O(n**2) check, could you consider a `group by` on the participant id to reduce the complexity here

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -0,0 +1,269 @@
+                sample for sample in samples if sample.participant_id == participant.id
+            ]
+
+            # Create a new participant object with the samples for this participant
```
Do you need to copy, if so why?

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -0,0 +1,269 @@
+        # go through all samples and get the meta and do something with it
+        # if we want stats for samples by participant, move this logic up before participant and make use of
+        # the filtered samples to map to each participant
+        for sample in samples:
```
Could you move this to a separate function, and return some data structure that contains all these stats? There's a lot of logic in here that IMO could also be broken into separate function.

A small note, there's very little computational difference between these two, so consider that when this gets refactored:

```python
for sample in samples:
    if sample.condition_1:
        list1.append(sample)
    if sample.condition_12:
        list2.append(sample)

# AND
for sample in samples:
    if sample.condition_1:
        list1.append(sample)

for sample in samples:
    if sample.condition_12:
        list2.append(sample)
```

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -0,0 +1,269 @@
+        )
+
+        return {
+            'collection_to_process_end_time': collection_to_process_end_time,
```
Use a data structure here

illusional commented on test/test_ourdna_dashboard.py:
```diff
@@ -0,0 +1,397 @@
+        self.pl = ParticipantLayer(self.connection)
+
+        participants = await self.pl.upsert_participants(
+            [
```
Can you move these static variables to be defined as constants to make it easier to scan the setup function

illusional commented on test/test_ourdna_dashboard.py:
```diff
@@ -0,0 +1,397 @@
+
+        self.participants_external_objects = [participant.to_external() for participant in participants]
+
+        samples_data = [
```
Why not construct the objects here directly rather than matching keys you need to select for, and instead keep them nested in the `ParticipantUpsert` (under the samples attribute) that will do all this inserting for you, eg: https://github.com/populationgenomics/metamist/blob/392d08191a79fbdeb2ac10f6a36c058111a007dc/test/test_graphql.py#L27-L31

illusional commented on db/python/layers/participant.py:
```diff
@@ -981,3 +981,9 @@ async def update_participant_family(
+    async def get_participants_and_samples_meta_by_project(
+        self, project: ProjectId
+    ) -> list[tuple[int, dict, dict]]:
+        """Get participants who have consented but not collected samples"""
```
This docstring doesn't match

illusional commented on models/models/ourdna.py:
```diff
@@ -0,0 +1,73 @@
+            participants_signed_not_consented=participants_signed_not_consented,
+        )
+
+    def to_dict(self) -> dict:
```
Can we return the actual object to GraphQL instead of to a generic dict, that gives you much better typing, plus you can use the experimental pydantic operator: https://strawberry.rocks/docs/integrations/pydantic.

That will give you a strongly types graphql query when you run `npm run compile`, and make the tests a bit more concrete

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -95,129 +101,231 @@ def get_collection_to_process_start_time(self, sample: Sample) -> int | None:
+        Fetches a key from the given meta, if it exists, and checks if it is not False or None
+        """
+        if isinstance(meta, str):
+            meta = json.loads(meta)
```
Does this happen often? Feels performant dangerous to potentially always deserialise a JSON string

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -95,129 +101,231 @@ def get_collection_to_process_start_time(self, sample: Sample) -> int | None:
+                value = meta.get(key)
+                if value is not False and value is not None:
+                    return True
+        return False
```
Do you want to throw an exception if the type isn't recognised?

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -95,129 +101,231 @@ def get_collection_to_process_start_time(self, sample: Sample) -> int | None:
-        check_project_ids: bool = True,
-        project_id: int = None,
-    ) -> dict:
+        project_id: ProjectId = None,
```
I think project_id here should be required, because it's used for queries below

illusional commented on models/models/ourdna.py:
```diff
@@ -0,0 +1,73 @@
+    participants_signed_not_consented: list[int] = []
+
+    @staticmethod
+    def from_sample(d: dict) -> 'OurDNADashboard':
```
I'm not really understanding the method name.
Is this method required, it obscures the names of the variables, where you can construct directly, and let pydantic do that heavy lifting for you. Pydantic will also throw an error if the types aren't right which I think is preferred behaviour.

If there are specific transformations you need to do here (it doesn't look like it), then for sure, but I'd collect with `**kwargs`, and only modify the fields you need modifying. Maybe worth us chatting about this when we catch-up!

illusional commented on test/test_ourdna_dashboard.py:
```diff
@@ -39,124 +39,112 @@ async def setUp(self) -> None:
         """Test get_dashboard"""
         sample_filter = SampleFilter(project=GenericFilter(eq=self.project_id))
-        dashboard = await self.odd.query(sample_filter)
+        dashboard = (await self.odd.query(sample_filter)).to_dict()
```
Following on from broken conversation above, not running to_dict here means the static type linting is as useful as running the test.

illusional commented on test/test_ourdna_dashboard.py:
```diff
@@ -0,0 +1,412 @@
+        assert total_samples_by_collection_event_name
+        assert isinstance(total_samples_by_collection_event_name, dict)
+
+        sample_tally: dict[str, int] = OrderedDict()
```
This is a test, so I don't thing it matters too much, but I think you could simplify this by using a `defaultdict` here, and then using unittests `self.assertDictEqual`, rather than a straight assert.

In tests, use assert when you want to do something with the typing, otherwise generally encourage using the [unittests methods](https://docs.python.org/3/library/unittest.html#assert-methods) as it provides pretty good support for telling you the value, vs differences are.

nevoodoo replied:
Great suggestion! I've ported most of the references where I used `assert` except where I need it for typing.

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -0,0 +1,371 @@
+        I want to know how long it took between blood collection and sample processing - SAMPLE TABLE
+        @fields: collection-time, process-end-time
+        """
+        _collection_time = self.get_meta_property(
```
I think this logic would be become simpler if there's a pre-processing step that pulls out these fields as pre-parsed dates, and then these values can be methods on that class. Pulling various tags from meta continuously, and parsing the dates each time feels a little tedious.

Maybe run a pre-processing step to get:

```python
class SampleProcessMeta:
    collection_time: datetime | None
    process_site: str
    process_start_time: datetime | None
    process_end_time: datetime | None

    @staticmethod
    def try_parse_datestr(datestr):
        try:
            return datetime.strptime(
                datestr, '%Y-%m-%d %H:%M:%S'
            )
        except:
            return None

    @staticmethod
    def from_sample(sample: Sample):
        collection_time_str = get_meta_property(sample=sample, property_name='collection-time')
        try:
        
        # ...
        return SampleProcessMeta(...)

    def get_collection_to_process_end_time_seconds(self) -> int:
        if not self.collection_time or not self.process_end_time:
            return None
        return int(self.process_end_time - self.collection_time)
```

What do you think?

illusional reviewed (commented):
Hey @nevoodoo, this is super close! On looking again, I think there's a way to simplify the logic of the layer, and nicely encapsulate that logic.

Could you also switch to using [unittest's assert methods](https://docs.python.org/3/library/unittest.html#assert-methods) in the test, they give better logging of values when something isn't equal - and you can use dicts without requiring them to be ordered (with dictEqual, eg)

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -15,93 +15,106 @@
-
-        self.sample_layer = SampleLayer(connection)
-        self.participant_layer = ParticipantLayer(connection)
+    meta: dict
```
Hmm, doesn't this end up as a static variable on the class? It's a different pattern to what we usually do. Can you just make this a regular class that gets constructed on input?

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -15,93 +15,106 @@
+        instance.meta = sample.meta
+        return instance
+
+    @property
```
Can use cached_property to reduce boilerplate
```suggestion
    @cached_property
```

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -15,93 +15,106 @@
+
+    @property
+    def get_lost_sample_properties(self) -> dict[str, Any]:
+        """Returns the properties to report for a sample that has been lost"""
```
```suggestion
        """Returns the normalised properties to report for a sample that has been lost"""
```

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -15,93 +15,106 @@
+    @staticmethod
+    def from_sample(sample: Sample) -> 'SampleProcessMeta':
+        """Create a SampleProcessMeta instance from a Sample instance."""
+        instance = SampleProcessMeta()
```
EG: it should be as simple as this
```suggestion
        instance = SampleProcessMeta(meta=sample.meta)
```

Or just change this whole thing to an init:

```python
def __init__(self, sample: Sample):
    self.meta = sample.meta
```

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -273,71 +289,27 @@ def process_samples_lost_after_collection(
-            ):
+            processed_meta = SampleProcessMeta.from_sample(sample)
+            time_to_process_start = processed_meta.collection_to_process_start_time
+            if time_to_process_start and time_to_process_start > 72 * 60 * 60:
```
Put this logic into the SampleProcessMeta, as `is_lost`

illusional commented on db/python/layers/ourdna/dashboard.py:
```diff
@@ -15,93 +15,106 @@
+        return self._collection_to_process_start_time
+
+    @property
+    def get_lost_sample_properties(self) -> dict[str, Any]:
```
If you store a reference to the sample, you can return a `OurDNALostSample` and get init checking through linting, make your life easy here!

illusional reviewed (changes requested):
Hey @nevoodoo, this is really close! Great that more logic is encapsulated in that class. 

Though, you can encode more logic and reduce the amount of key accesses here. I've less direct in earlier reviews, but I'll be more prescriptive, encode those key accesses as actual fields. 

- You're already encoding the field names as strings, simplify the amount of string constants here!
- we're parsing dates every time we want to use it. In the below equations, we're parsing everything twice.
- there's no error handling here for bad date formats (want to add a test for this?)
- You can unit test this class without inserting anything into the database


Here's my direct suggestion:

```python
@staticmethod
def try_parse_datetime(d) -> datetime | None:
    try:
        return datetime.strptime(d, '%Y-%m-%d %H:%M:%S')
    except:
        return None

@cached_property
def process_start_time(self) -> datetime | None:
    return self.try_parse_datetime(self.get_property('process-start-time'))

@cached_property
def processing_time(self) -> tuple[str | None, int | None]:
    """Get processing times and site for a sample."""
    if not (self.process_start_time and self.process_end_time):
        return None
    return (self.process_end_time - self.process_start_time).total_seconds

@cached_property
def processing_time_by_site(self) -> tuple[str | none, int | None]:
    return self.get_property('processing-site'), self.processing_time

illusional reviewed (approved):
Amazing, thanks for actioning my annoyingly progressive feedback!
